### ISSUE TITLE
YAML Front Matter support in Markdown

### ISSUE THREAD

**Issue by jackmaney** [Glavin001/atom-beautify#169]
So, I have the following markdown with [YAML front matter (for use with Jekyll)](http://jekyllrb.com/docs/frontmatter/):

```
---
title: This is a title!
---

stuff

more stuff
```

When I beautify this, the front matter disappears:

```
stuff

more stuff
```

Adding `GitHub Markdown` to the disabled languages prevents the front matter from being eaten. However, it would be nice to have the beautification work.

Thank you for your time.

**Comment by Glavin001** [comment]
There are currently no beautifiers (that I have found) that support YAML Front Matter.
Possible solutions: detect that the file contents begins with YAML Front Matter, and if so, remove it from contents, beautify contents, and then prepend the YAML front matter to the post-beautified contents. This should work well.
No ETA on this functionality right now. Thanks for making this issue!

**Comment by Glavin001** [comment]
- [ ] Add package option for enabling YAML Front Matter
- If YAML Front Matter is enabled, on Markdown beautification, 
  - [ ] first process with https://github.com/dworthen/js-yaml-front-matter and then take the `__content` and pass that to the Markdown beautifier
  - [ ] take the result of Markdown beautifier of `__content` and concat the YAML Front Matter to the top and return that result

**Comment by Glavin001** [comment]
I started working on this however I realize now that the `cli-beautify` base class is not very customizable and does not meet the requirements I have for this feature. I think completing #164 should come first as it will force me to redesign how CLI-beautifiers are handled and now I can start brainstorming an improved API for supporting existing and adding more CLI-beautifiers in the future.

**Comment by Glavin001** [comment]
I will definitely update this after #164, however it is at least working now and you can use it while I develop out #164.

### Before

```markdown
---

title: This is a title!

name: Derek Worthen
age: young
contact:
email: email@domain.com
address: some location
pets:
- cat
- dog
- bat
match: !!js/regexp /pattern/gim
run: !!js/function function() { }



---

- item
-  item
-   item

1. one
2.  two
2.   three
```

### After

```markdown
---
title: "This is a title!"
name: Derek Worthen
age: young
contact: null
email: "email@domain.com"
address: some location
pets:
  - cat
  - dog
  - bat
match: !<tag:yaml.org,2002:js/regexp> /pattern/gmi
run: !<tag:yaml.org,2002:js/function> "function () {\n   \n}"
---
-   item
-   item
-   item

1.  one
2.  two
3.  three
```

Hope that looks good to you!

**Comment by Glavin001** [comment]
The build is failing for release v0.20.0.
See https://travis-ci.org/Glavin001/atom-beautify/builds/45655633
I think this is because of https://github.com/dworthen/js-yaml-front-matter/issues/2
Until that issue has been resolved, I may have to disable / remove this feature. @jackmaney could you please update and check for yourself if it is working? Thanks.

**Comment by Glavin001** [comment]
Fortunately it looks like the YAML Front Matter library wraps `js-yaml`, so I can remove `js-yaml` and use `yaml-front-matter` instead.

**Comment by Glavin001** [comment]
Published to v0.20.1! Enjoy :smiley:

**Comment by jackmaney** [comment]
Awesome. Thank you so much! :D

**Comment by jackmaney** [comment]
Just updated to v0.20.1, and I'm seeing this:

# Before

```
---

title: This is a title!

name: Derek Worthen
age: young
contact:
email: email@domain.com
address: some location
pets:
- cat
- dog
- bat
match: !!js/regexp /pattern/gim
run: !!js/function function() { }



---

- item
-  item
-   item

1. one
2.  two
2.   three
```

# After

```
------------------------------------------------------------------------

title: This is a title!

name: Derek Worthen age: young contact: email: email@domain.com address:
some location pets: - cat - dog - bat match: !!js/regexp /pattern/gim
run: !!js/function function() { }

------------------------------------------------------------------------

-   item
-   item
-   item

1.  one
2.  two
3.  three

```

**Comment by jackmaney** [comment]
Hmmm, strangely enough, the same exact `before` and `after` in the opening comment is still occurring:

# Before

```
---
title: This is a title!
---

stuff

more stuff
```

# After

```
stuff

more stuff

```

**Comment by Glavin001** [comment]
That's very strange. Travis CI already OKed the unit test for the more complex case. I'll have to take another look once I'm home. 

—

**Comment by Glavin001** [comment]
I'm home now. I am still receiving correct results for both of those tests. Could you run command `Window: Reload` and / or uninstall and reinstall Atom Beautify?
